docs/src/lens.md

```diff
@@ -1,149 +1,251 @@
+
+impl Lens<Container2, Name> for NameLens {
+    // .. identical to previous example
+}
 ```
 
-Here `In` refers to the input to the `Lens` and `Out` is the output. `F` is a
-closure that can return a result, `R`.
+As you'll see, we've introduced `Rc`: the reference-counted pointer. You will see this and its multithreaded cousin `Arc` used pervasively in the examples. Now, the only time we actually have to copy memory is when `Rc::make_mut` is called in the `f` in `with_mut`. This means that in the case where nothing changes, all we will be doing is incrementing and decrementing reference counts. Moreover, we give the compiler the opportunity to inline `f` and `with`/`with_mut`, making this abstraction potentially zero-cost (disclaimer: I haven't actually studied the produced assembly to validate this claim).
 
-Now, instead of just being passed `Out` directly from the function, we pass the
-function a closure that will *itself* be passed an `Out`; if our closure returns
-a result, that will be given back to us.
+The trade-off is that we introduce more complexity into the `Name` type: to make changes to the data we have to use `Rc::make_mut` to get mutable access to the `String`. (The code in the lens will ensure that the newer copy of the `Rc`d data is saved to the outer type.) This means the writing fast druid code requires knowledge of the Rust pointer types (`Rc`/`Arc`, and also potentially `RefCell`/`Mutex`).
 
-This is unnecessary in the case of non-mutable access, but it is important for
-mutable access, because in many circumstances (such as when using an `Rc` or
-`Arc`) accessing a field mutably is expensive even if you don't do any mutation.
+We can actually do even better than this. Suppose that we are working on a vector of data rather than a string. We can import the `im` crate to get collections that use *structural sharing*, meaning that even when the vector is mutated, we only *Clone* what we need to. Because `im` is so useful, it is included in `druid` (behind the `im` feature).
 
-In any case, the real implementation of our lens would look like,
+```rust
+struct Container2 {
+    // Pretend that it's the 1980s and we store only ASCII names.
+    first_name: im::Vector<u8>,
+    last_name: im::Vector<u8>,
+    age: u16,
+}
+
+struct Name {
+    first: im::Vector<u8>,
+    last: im::Vector<u8>,
+}
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:completed_lens_real}}
+struct NameLens;
+
+impl Lens<Container2, Name> for NameLens {
+    // .. identical to previous example
+}
 ```
 
-That seems pretty simple and fairly annoying to write, which is why you
-generally don't have to.
+Now in addition to almost free `Clone`s, we also have cheap incremental updates to the data itself. In the case of names, this isn't that important, but if the vector had `1_000_000_000` elements, we could still make changes in only *O(log(n))* time (in this case the difference between `1_000_000_000` and `30` - pretty big!).
```

Comment:
I found this a little confusing - where does 30 come from?
I quickly checked for both the natural log and the base-10 log, and neither was 30 - maybe a typo?
```
ln(1E9)  = 20.72
log(1E9) = 9.00
```

Otherwise, this is an excellent point! It's a huuuuge savings on even moderately sized collections.

Reply:
It was from log2, it was really meant to be illustrative only - in reality you probably do better by having 32 elements per node (or even more).